WIP: FEATURE: Refactoring for Neos 9

mficzel opened the pull request:
The flowQuery operations `referenceNodes()` and `backReferenceNodes` in combination to search for documents that have similar taxons assigned. 

```neosfusion
similarDocuments = ${
    q(documentNode)
        .referenceNodes('taxonomyReferences')                  // the taxons the current document references 
        .backReferenceNodes('taxonomyReferences')              // all nodes that reference one of the same taxons
        .filter('[instanceof Neos.Neos:Document]')             // only documents
        .remove(documentNode)                                  // but nut the current one
        .unique()                                              // every document only once
        .get()
    }  
```

The package now includes the following flowQuery operations:

- `referencedTaxonomies()`:  the taxons referenced by the documents in flowQuery context
- `referencingTaxonomies()`:  the documents referencing by the taxons in flowQuery context
- `subTaxonomies()` :  sub-taxons of taxons in the context
- `withSubTaxonomies()`:  current taxons in the context plus sub-taxons 

```neosfusion
similarDocuments = ${
  q(documentNode)
    .referencedTaxonomies()                                // the taxons the current document references
    .withSubTaxonomies()                                   // including all sub taxons
    .referencingTaxonomies()                               // the documents that reference the same taxons
    .remove(documentNode)                                  // but nut the current one
    .get()
  }   
```



Must have:
- [x] BE Module create, edit, delete, show Vocabularies 
- [x] BE Module create, edit, delete, show Taxonomies 
- [x] BE Module - Dimension switch
- [x] BE Module - alphabetical order
- [x] Command Controller
- [x] Taxonomy Editor with secondary editor
- [x] Update Documentation

Nice to have:
- BE Module - default node comparison
- BE Module - protect tethered nodes from deletion

For now the following two branches are needed for this to work:
- ~Neos https://github.com/neos/neos-development-collection/pull/4339~
- ~Neos UI https://github.com/neos/neos-ui/pull/3530~

mficzel committed:
WIP: Initial refactioring for neos 9

The backend mule is basically working again

mficzel committed:
TASK: Reimplement dimension switch in backend module

mficzel committed:
TASK: Minor cleanup

mficzel committed:
TASK: Add name property with aggregate scope to taxonomies and vocabularies

mficzel committed:
TASK: Move NodeTypes to specific folder and manage NodeNames for vocabularies and taxonomies

mficzel committed:
TASK: Cleanup and reimplement parts of the command controller

mficzel committed:
TASK: Adjust starting point and rename mixin to TaxonomyReferences to be in line with the property name

mficzel committed:
TASK: Adjust TaxonomyEditor to work with Neos 9

- Adjust route path to be in /neos namesapace to avoid conflicts with frontendlogin
- Adjust SecondaryInspectorController to new code and pass startingPoint and contextPath seperately
- Various cleanups

mficzel committed:
TASK: Cleanup and improve names

mficzel committed:
TASK: Add linting and static analysis

mficzel committed:
TASK: Cleanup

mficzel committed:
TASK: Make linter almost happy

as soon as the absolute path prs are merged all should become green

mficzel committed:
FEATURE: Add flowQueryOperations `taxonomyReferencingNodes`, `taxonomiesAncestors`, `taxonomyDescendants`, `taxonomyChildren`, `taxonomyVocabularies`

For now those operations are working on Nodes ... References support will be added later.

mficzel committed:
TASK: Make linter as possible for now

mficzel committed:
TASK: Taxonomy FlowQuery operations `taxonomyDescendants`, `taxonomyAncestors`, `taxonomyWithDescendants`, `taxonomyWithAncestors`

mficzel committed:
TASK: Make phpstan happy

mficzel committed:
TASK: Update docs and remove eel helper

mficzel committed:
TASK: Remove flow query operations, those should be added in a separate package

ahaeslich commented on Classes/Controller/ModuleController.php:
```diff
@@ -46,477 +56,418 @@ class ModuleController extends ActionController
-        $nodeTemplate->setNodeType($taxonomyNodeType);
-        $nodeTemplate->setName(CrUtitlity::renderValidNodeName($properties['title']));
+        $parentNode = $this->taxonomyService->getNodeByNodeAddress($parentNodeAddress);
+        $vocabularyNode = $this->taxonomyService->findVocabularyForNode($parentNode);
```
This will fail if `$parentNode` is already a vocabulary.

```suggestion        
        $vocabularyNode = $this->taxonomyService->getTaxonomyNodeTypeName()->equals($parentNode->nodeTypeName)
            ? $this->taxonomyService->findVocabularyForNode($parentNode)
            : $parentNode;
```

ahaeslich replied:
Will check this again as I didn't need this while using the current neos dev collection

mficzel replied:
Confirmed and adjusted. Thanks for reporting

ahaeslich committed:
TASK: Use named `create` constructor for all commands

Relates: https://github.com/neos/neos-development-collection/pull/4489

mficzel committed:
Merge pull request #67 from queoGmbH/create-command-constructor

TASK: Use named create constructor for all commands

mficzel committed:
TASK: Add flowQuery operations

- `referencedTaxonomies()`:  the taxons referenced by the documents in flowQuery context
- `referencingTaxonomies()`:  the documents referencing by the taxons in flowQuery context
- `subTaxonomies()` :  sub-taxons of taxons in the context
- `withSubTaxonomies()`:  current taxons in the context plus sub-taxons

mficzel commented:
@ahaeslich i am usure about the names and interfaces of the flowQuery operations

- `referencedTaxonomies()`: Better naming and is there a need to specify the reference name and taxon type?
- `referencingTaxonomies()`: Better naming and is there a need to specify the reference name and document type?
- `subTaxonomies()`: Is there a need to specify the depth taxon type ?
- `withSubTaxonomies()`: Os there a need to specify the depth taxon type ?

mficzel committed:
TASK: Use cr registry to get the subgraph and also fix bug when creating taxons directly below the vocabulary node

mficzel committed:
TASK: Remove Silhouettes from configuration

mficzel committed:
TASK: Avoid using node->nodeType->isOf ... instead use node->nodeTypeName->equals

This is not exactly the same as the new solution includes subtypes but the isOf method requires the nodeTypeManager which is expensive.

mficzel committed:
TASK: Remove configurable nodeTypeNames and hardwire them in TaxonomyService

ahaeslich commented on Classes/Service/TaxonomyService.php:
```diff
@@ -1,225 +1,252 @@
+use Neos\ContentRepository\Core\SharedModel\Workspace\WorkspaceName;
+use Neos\ContentRepositoryRegistry\ContentRepositoryRegistry;
+use Neos\Flow\Annotations as Flow;
+use Neos\Neos\Domain\Exception\LiveWorkspaceIsMissing;
```
@mficzel I just found that this one got deleted with https://github.com/neos/neos-development-collection/pull/4563/commits/144ad066dbe6555da5fe5b139cd991cb77082399

mficzel replied:
adjusted

mficzel committed:
TASK: Remove usage of upstream deleted exception

mficzel committed:
TASK: Rebase current user workspace after changes to taxonomies

Commit:
Properly handle tethered nodes

also properly find reference vocabulary node

ahaeslich committed:
TASK: Adjust to renamed NodeTypeConstraints

Relates: neos/neos-development-collection#4522

mficzel committed:
Merge pull request #69 from queoGmbH/task/adjustToRenamedNodeType

TASK: Adjust to renamed NodeTypeConstraints

Commit:
Merge branch 'neos-9' into tetheredNodes

Commit:
Remove redundant StopCommandException

nezaniel committed:
Merge pull request #68 from sitegeist/tetheredNodes

Properly handle tethered nodes

grebaldi committed:
TASK: Remove plow-js, Update to 8.3 extensibility and esbuild stack

nezaniel committed:
Merge pull request #73 from grebaldi/bugfix/neos-9-ui-compatibility

BUGFIX: Adjust to Neos 9.0 UI changes
